File: windows-winui/Core/Archive.cs
using System.IO;

namespace SiftWinUI.Core;

/// Claude Code deletes transcripts older than `cleanupPeriodDays` (30 by default), which
/// means the thing this app exists to search quietly disappears from under it. Sift keeps
/// its own copy: every transcript it indexes is archived, and a session whose original has
/// been cleaned up stays searchable and readable.
///
/// The archive is a plain tree of the original .jsonl files, so it needs no migration and
/// can be read by anything, including Claude Code itself once a file is put back.
public static class Archive
{
    public static string Root => Path.Combine(AppPaths.SupportDir, "archive");

    public static string PathFor(string projectId, string sessionId) =>
        Path.Combine(Root, Sanitise(projectId), Sanitise(sessionId) + ".jsonl");

    /// Copies a transcript in, unless an identical copy is already there. Returns true when
    /// something was written.
    public static bool Keep(string sourcePath, string projectId, string sessionId)
    {
        var target = PathFor(projectId, sessionId);
        try
        {
            var source = new FileInfo(sourcePath);
            if (!source.Exists) return false;
            var existing = new FileInfo(target);
            if (existing.Exists && existing.Length == source.Length &&
                existing.LastWriteTimeUtc >= source.LastWriteTimeUtc) return false;

            Directory.CreateDirectory(Path.GetDirectoryName(target)!);
            File.Copy(sourcePath, target, overwrite: true);
            return true;
        }
        catch (IOException) { return false; }
        catch (UnauthorizedAccessException) { return false; }
    }

    public static bool Has(string projectId, string sessionId) => File.Exists(PathFor(projectId, sessionId));

    /// Puts a transcript back where Claude Code expects it, so `claude --resume` can find a
    /// session that was cleaned up. Returns the restored path, or null if there was nothing
    /// to restore.
    public static string? Restore(string projectId, string sessionId)
    {
        var archived = PathFor(projectId, sessionId);
        if (!File.Exists(archived)) return null;
        try
        {
            var target = Path.Combine(AppPaths.ProjectsRoot, projectId, sessionId + ".jsonl");
            Directory.CreateDirectory(Path.GetDirectoryName(target)!);
            if (!File.Exists(target)) File.Copy(archived, target);
            return target;
        }
        catch (IOException) { return null; }
        catch (UnauthorizedAccessException) { return null; }
    }

    public static (int Files, long Bytes) Size()
    {
        if (!Directory.Exists(Root)) return (0, 0);
        var files = Directory.EnumerateFiles(Root, "*.jsonl", SearchOption.AllDirectories).ToList();
        long bytes = 0;
        foreach (var f in files)
        {
            try { bytes += new FileInfo(f).Length; } catch (IOException) { }
        }
        return (files.Count, bytes);
    }

    /// Directory names come from Claude Code and session ids from filenames, but a path
    /// separator slipping through would write outside the archive.
    private static string Sanitise(string name)
    {
        foreach (var c in Path.GetInvalidFileNameChars()) name = name.Replace(c, '-');
        return name.Replace("..", "-");
    }
}

/// Claude Code's own retention, in `~/.claude/settings.json`. Turning it off is what stops
/// the deletion happening in the first place; the archive is what covers everything that
/// was already lost or gets removed by another machine.
public static class ClaudeRetention
{
    public static string SettingsPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".claude", "settings.json");

    public const int Forever = 36500;   // a hundred years, the largest value that stays sane

    public static int? CurrentDays()
    {
        try
        {
            if (!File.Exists(SettingsPath)) return 30;   // Claude Code's own default
            using var doc = System.Text.Json.JsonDocument.Parse(File.ReadAllText(SettingsPath));
            return doc.RootElement.TryGetProperty("cleanupPeriodDays", out var v) && v.TryGetInt32(out var days)
                ? days : 30;
        }
        catch { return null; }
    }

    /// Rewrites only `cleanupPeriodDays`, leaving every other setting exactly as it was.
    public static bool SetDays(int days)
    {
        try
        {
            var dir = Path.GetDirectoryName(SettingsPath)!;
            Directory.CreateDirectory(dir);
            var node = File.Exists(SettingsPath)
                ? System.Text.Json.Nodes.JsonNode.Parse(File.ReadAllText(SettingsPath)) as System.Text.Json.Nodes.JsonObject
                : new System.Text.Json.Nodes.JsonObject();
            node ??= new System.Text.Json.Nodes.JsonObject();

            if (File.Exists(SettingsPath))
                File.Copy(SettingsPath, SettingsPath + ".sift-backup", overwrite: true);

            node["cleanupPeriodDays"] = days;
            File.WriteAllText(SettingsPath, node.ToJsonString(
                new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));
            return true;
        }
        catch { return false; }
    }
}

File: web/lib/archive.mjs
import { copyFile, mkdir, stat, readFile, writeFile, readdir } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { homedir } from 'node:os';

/// Claude Code deletes transcripts past `cleanupPeriodDays` (30 by default), which means
/// the corpus this tool searches quietly disappears from under it. Sift keeps its own copy
/// of everything it indexes, so a cleaned-up session stays searchable and readable.
///
/// The archive is a plain tree of the original .jsonl files: no format to migrate, and a
/// file can be put straight back where Claude Code expects it.
export function archiveRoot(supportDir) {
  return join(supportDir, 'archive');
}

/// Directory names come from Claude Code and session ids from filenames; a separator
/// slipping through would write outside the archive.
function sanitise(name) {
  return name.replace(/[<>:"|?*\\/]/g, '-').replace(/\.\./g, '-');
}

export function archivePathFor(supportDir, projectId, sessionId) {
  return join(archiveRoot(supportDir), sanitise(projectId), sanitise(sessionId) + '.jsonl');
}

/// Copies a transcript in unless an identical copy is already there.
export async function keep(supportDir, sourcePath, projectId, sessionId) {
  const target = archivePathFor(supportDir, projectId, sessionId);
  try {
    const source = await stat(sourcePath);
    try {
      const existing = await stat(target);
      if (existing.size === source.size && existing.mtimeMs >= source.mtimeMs) return false;
    } catch { /* not archived yet */ }
    await mkdir(dirname(target), { recursive: true });
    await copyFile(sourcePath, target);
    return true;
  } catch {
    return false;
  }
}

/// Puts a transcript back where Claude Code expects it, so `claude --resume` can find a
/// session that was cleaned up.
export async function restore(supportDir, projectsRoot, projectId, sessionId) {
  const archived = archivePathFor(supportDir, projectId, sessionId);
  if (!existsSync(archived)) return null;
  const target = join(projectsRoot, projectId, sessionId + '.jsonl');
  if (existsSync(target)) return target;
  try {
    await mkdir(dirname(target), { recursive: true });
    await copyFile(archived, target);
    return target;
  } catch {
    return null;
  }
}

export async function archiveSize(supportDir) {
  const root = archiveRoot(supportDir);
  let files = 0;
  let bytes = 0;
  try {
    for (const dir of await readdir(root, { withFileTypes: true })) {
      if (!dir.isDirectory()) continue;
      for (const name of await readdir(join(root, dir.name))) {
        if (!name.endsWith('.jsonl')) continue;
        files += 1;
        try { bytes += (await stat(join(root, dir.name, name))).size; } catch { /* raced */ }
      }
    }
  } catch { /* nothing archived yet */ }
  return { files, bytes };
}

/// Claude Code's own retention lives in ~/.claude/settings.json. Turning it off is what
/// stops the deletion happening at all; the archive covers what other machines remove.
export const settingsPath = () => join(homedir(), '.claude', 'settings.json');

export async function currentRetentionDays() {
  try {
    const value = JSON.parse(await readFile(settingsPath(), 'utf8')).cleanupPeriodDays;
    return Number.isInteger(value) ? value : 30;
  } catch {
    return 30;   // Claude Code's own default when unset
  }
}

export const FOREVER = 36500;

/// Rewrites only `cleanupPeriodDays`, keeping every other setting, after a backup.
export async function setRetentionDays(days) {
  const path = settingsPath();
  let settings = {};
  try {
    const text = await readFile(path, 'utf8');
    settings = JSON.parse(text);
    await writeFile(path + '.sift-backup', text);
  } catch { /* no settings file yet */ }
  settings.cleanupPeriodDays = days;
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, JSON.stringify(settings, null, 2) + '\n');
  return days;
}
